vms/platformvm/validators/manager.go

```diff
@@ -198,146 +225,162 @@ func (m *manager) getValidatorSetFrom(
+		ctx,
+		validatorSet,
+		currentHeight,
+		lastDiffHeight,
+		constants.PlatformChainID,
+	)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	err = m.state.ApplyValidatorPublicKeyDiffs(
+		ctx,
+		validatorSet,
+		currentHeight,
+		lastDiffHeight,
+	)
+	return validatorSet, currentHeight, err
+}
+
+func (m *manager) getCurrentPrimaryValidatorSet(
+	ctx context.Context,
+) (map[ids.NodeID]*validators.GetValidatorOutput, uint64, error) {
 	currentValidators, ok := m.cfg.Validators.Get(constants.PrimaryNetworkID)
 	if !ok {
 		// This should never happen
 		m.log.Error(ErrMissingValidatorSet.Error(),
 			zap.Stringer("subnetID", constants.PrimaryNetworkID),
 		)
-		return nil, ErrMissingValidatorSet
+		return nil, 0, ErrMissingValidatorSet
 	}
 
-	// Node ID --> Validator information for the node validating the Primary
-	// Network.
-	validatorSet := currentValidators.Map()
-
-	// Rebuild primary network validators at [height]
-	for diffHeight := currentHeight; diffHeight > targetHeight; diffHeight-- {
-		weightDiffs, err := m.state.GetValidatorWeightDiffs(diffHeight, constants.PlatformChainID)
-		if err != nil {
-			return nil, err
-		}
-		for nodeID, weightDiff := range weightDiffs {
-			if err := applyWeightDiff(validatorSet, nodeID, weightDiff); err != nil {
-				return nil, err
-			}
-		}
-
-		pkDiffs, err := m.state.GetValidatorPublicKeyDiffs(diffHeight)
-		if err != nil {
-			return nil, err
-		}
-		for nodeID, pk := range pkDiffs {
-			if vdr, ok := validatorSet[nodeID]; ok {
-				// The validator's public key was removed at this block, so it
-				// was in the validator set before.
-				vdr.PublicKey = pk
-			}
-		}
-	}
-	return validatorSet, nil
+	currentHeight, err := m.getCurrentHeight(ctx)
+	return currentValidators.Map(), currentHeight, err
 }
 
 func (m *manager) makeSubnetValidatorSet(
-	currentHeight uint64,
+	ctx context.Context,
 	targetHeight uint64,
 	subnetID ids.ID,
-) (map[ids.NodeID]*validators.GetValidatorOutput, error) {
-	currentValidators, ok := m.cfg.Validators.Get(subnetID)
-	if !ok {
-		currentValidators = validators.NewSet()
-		if err := m.state.ValidatorSet(subnetID, currentValidators); err != nil {
-			return nil, err
-		}
+) (map[ids.NodeID]*validators.GetValidatorOutput, uint64, error) {
+	subnetValidatorSet, primaryValidatorSet, currentHeight, err := m.getCurrentValidatorSets(ctx, subnetID)
+	if err != nil {
+		return nil, 0, err
 	}
-
-	// Node ID --> Validator information for the node validating the Subnet.
-	subnetValidatorSet := currentValidators.Map()
-
-	// Rebuild subnet validators at [targetHeight]
-	for diffHeight := currentHeight; diffHeight > targetHeight; diffHeight-- {
-		weightDiffs, err := m.state.GetValidatorWeightDiffs(diffHeight, subnetID)
-		if err != nil {
-			return nil, err
-		}
-
-		for nodeID, weightDiff := range weightDiffs {
-			if err := applyWeightDiff(subnetValidatorSet, nodeID, weightDiff); err != nil {
-				return nil, err
-			}
-		}
+	if currentHeight < targetHeight {
```

Comment:
How could this happen? (or is it meant just to protect from bad inputs)

Reply:
This can happen if someone requests the validator set for a P-chain height that hasn't happened yet. (So that would be an invalid input)